src/mini_s3.erl

```diff
@@ -448,472 +290,141 @@ if_not_empty("", _V) ->
+s3_url(Method, BucketName0, Key0, Lifetime, RawHeaders, Config)
+  when is_list(BucketName0), is_list(Key0), is_tuple(Config) ->
+    [BucketName, Key] = [ms3_http:url_encode_loose(X) || X <- [BucketName0, Key0]],
+    RequestURI = erlcloud_s3:make_presigned_v4_url(Lifetime, BucketName, Method, Key, [], RawHeaders, Config),
+    iolist_to_binary(RequestURI).
+
+-spec s3_url(atom(), string(), string(), non_neg_integer(), proplists:proplist(), string(), aws_config()) -> binary().
+s3_url(Method, BucketName0, Key0, Lifetime, RawHeaders, Date, Config)
+  when is_list(BucketName0), is_list(Key0), is_tuple(Config) ->
+    [BucketName, Key] = [ms3_http:url_encode_loose(X) || X <- [BucketName0, Key0]],
+    RequestURI = erlcloud_s3:make_presigned_v4_url(Lifetime, BucketName, Method, Key, [], RawHeaders, Date, Config),
+    iolist_to_binary(RequestURI).
+
+%-----------------------------------------------------------------------------------
+% implementation of expiration windows for sigv4, for making batches
+% of cacheable presigned URLs
+%
+%          past       present      future
+%                        |
+% ------+------+------+--+---+------+------+------+------
+%       |      |      |  |   |      |      |      |   time (exp-wins)
+% ------+------+------+--+---+------+------+------+------
+%                     |    ^ |
+%      x-amz-date ----+    | +---- x-amz-expires
+%                        |-|
+%                        TTL
+%                     |------|
+%                     Lifetime
+%
+% given a TTL, x-amz-expires should be set to be the closest expiry-window
+% boundary >= present+TTL, ie present+TTL selects the expiry-window.
+%
+% 1) segment all of time into 'windows' of width expiry-window-size.
+% 2) align x-amz-date to nearest expiry-window boundary less than present time.
+% 3) align x-amz-expires to nearest expiry-window boundary greater than present time.
+% 4) the right edge of present+TTL is a 'selector' to determine which expiration
+%    window we are in, thus determining final value of x-amz-expires and Lifetime.
+% 5) while x-amz-expires - present < TTL, x-amz-expires += expiry-window-size.
+% 6) Lifetime = x-amz-expires - x-amz-date.
+%-----------------------------------------------------------------------------------
+-spec make_expire_win(non_neg_integer(), non_neg_integer()) -> {string(), non_neg_integer()}.
+make_expire_win(TTL, ExpireWinSiz) when ExpireWinSiz > 0 ->
+    Present = calendar:datetime_to_gregorian_seconds(calendar:now_to_universal_time(os:timestamp())),
+    XAmzDateSec = Present div ExpireWinSiz * ExpireWinSiz,
+    ExpirWinMult = ((TTL div ExpireWinSiz) + (case TTL rem ExpireWinSiz > 0 of true -> 1; _ -> 0 end)),
+    XAmzExpires = case ExpirWinMult of 0 -> 1; _ -> ExpirWinMult end * ExpireWinSiz + XAmzDateSec,
+    Lifetime = XAmzExpires - XAmzDateSec,
+    {erlcloud_aws:iso_8601_basic_time(calendar:gregorian_seconds_to_datetime(XAmzDateSec)), Lifetime}.
```

Comment:
In the previous code we were concerned about the day boundary? Is that a concern anymore? I would have thought it is more of a concern because of how SigV4 handles the credential context.

Reply:
I tried to look into this somewhat.  I wish I understood the issue better (perhaps you can explain it to me).  And I wish I understood pre-existing code better.  What I can say is, my code does all calculations on 'seconds' (seconds since whenever timekeeping started haha) and has absolutely no concept about days or day boundaries.  At the end, it converts these seconds into a date/time.  So does a lack of a concept of days and day boundaries mean my code has a problem?  Or does it mean the opposite - that there can't be a problem because it has no concept of such things?  I don't know.

Perhaps we can look into this together.  I can explain to you how my code works, and draw up diagrams as necessary.  If you know how the previous code works, we can determine if there is an issue.  Perhaps you can also give me a 'day boundary' to run through my code, and we can see if it comes up with the correct calculation or not.

Reply:
One thing is, sigv4 presigned URLs cannot be valid for more than 1 week.  But my code probably happily allows you to create presigned URLs that are greater than one week.  I don't know if this is a problem.  My assumption was that we should simply not ask for presigned URLs which are valid for more than one week, i.e. don't specify a TTL or an expiration window of a week or greater.

Reply:
Thinking about the above more, it seems at the very least we could have a week boundary issue.  It might be possible to specify inputs that 'accidentally' give expirations greater than 1 week.  I can draw this out for you to see if you concur.  Perhaps I can add code to simply curb or squelch any expiration going over a week to a week.

Reply:
Added squelching of output URL lifetimes to 1 week.

Reply:
```
% specified TTL of greater than 1 week (604800 seconds)
% output lifetime is squelched to 1 week

> mini_s3:make_expire_win(999999,1).
{"20200916T215530Z",604800}

% specified expiration window of greater than 1 week (604800 seconds)
% output lifetime is squelched to 1 week

9> mini_s3:make_expire_win(1,999999).
{"20200910T055033Z",604800}
```

Reply:
Also, just checked.  It seems I reject presigned URLs with expirations greater than 1 week in bookshelf authentication (bksw_sec.erl).

Reply:
See some analysis further down on this.

Reply:
See some analysis further down on this.

Reply:
Looks like the day boundary that was being handles was for the purpose of simplifying the expiration times for the user, it seems to not have any effect on caching. Looking at https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html, It seems like as long as the x-amz-date header value (which is the time the request was made)  is the same as the date component in the credential scope, the authentication should be successful. So keeping the date boundary would be good to have to not break user expectations.

Reply:
That being said, @lbakerchef the Lifetime returned in the make_expire_win function did not match what I expected to see in the case where TTL = 60 mins and Interval = 15 mins. Would like to understand if I am missing something here.

Reply:
@PrajaktaPurohit is this conversation resolved since you added-in the old expiry windows tests?  Or should I leave this open?

Reply:
Reminder: Add @PrajaktaPurohit 's contribution to this branch's code.